File: bindings/swift/Sources/UMICP/Discovery/ServerInfo.swift
import Foundation

/// Server information for tool discovery
public struct ServerInfo: Codable, Equatable {
    public let server: String
    public let version: String
    public let `protocol`: String
    public var features: [String]?
    public var operationsCount: Int?
    public var mcpCompatible: Bool?
    public var metadata: [String: AnyCodable]?

    enum CodingKeys: String, CodingKey {
        case server, version
        case `protocol` = "protocol"
        case features
        case operationsCount = "operations_count"
        case mcpCompatible = "mcp_compatible"
        case metadata
    }

    public init(
        server: String,
        version: String,
        protocol: String,
        features: [String]? = nil,
        operationsCount: Int? = nil,
        mcpCompatible: Bool? = nil,
        metadata: [String: AnyCodable]? = nil
    ) {
        self.server = server
        self.version = version
        self.protocol = `protocol`
        self.features = features
        self.operationsCount = operationsCount
        self.mcpCompatible = mcpCompatible
        self.metadata = metadata
    }

    /// Builder pattern for features
    public func withFeatures(_ features: [String]) -> ServerInfo {
        var copy = self
        copy.features = features
        return copy
    }

    /// Builder pattern for operations count
    public func withOperationsCount(_ count: Int) -> ServerInfo {
        var copy = self
        copy.operationsCount = count
        return copy
    }

    /// Builder pattern for MCP compatibility
    public func withMcpCompatible(_ compatible: Bool) -> ServerInfo {
        var copy = self
        copy.mcpCompatible = compatible
        return copy
    }

    /// Builder pattern for metadata
    public func withMetadata(_ metadata: [String: AnyCodable]) -> ServerInfo {
        var copy = self
        copy.metadata = metadata
        return copy
    }

    /// Convert to dictionary representation
    public func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "server": server,
            "version": version,
            "protocol": `protocol`
        ]

        if let features = features {
            dict["features"] = features
        }
        if let operationsCount = operationsCount {
            dict["operations_count"] = operationsCount
        }
        if let mcpCompatible = mcpCompatible {
            dict["mcp_compatible"] = mcpCompatible
        }
        if let metadata = metadata {
            dict["metadata"] = metadata.mapValues { $0.value }
        }

        return dict
    }
}

File: bindings/csharp/Umicp.Core/ToolDiscovery/ServerInfo.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Umicp.Core.ToolDiscovery;

/// <summary>
/// Server information for discovery
/// </summary>
public class ServerInfo
{
    /// <summary>
    /// Server name/identifier
    /// </summary>
    [JsonPropertyName("server")]
    public string Server { get; set; } = string.Empty;

    /// <summary>
    /// Server version
    /// </summary>
    [JsonPropertyName("version")]
    public string Version { get; set; } = string.Empty;

    /// <summary>
    /// Protocol version (e.g., "UMICP/0.2")
    /// </summary>
    [JsonPropertyName("protocol")]
    public string Protocol { get; set; } = string.Empty;

    /// <summary>
    /// List of supported features
    /// </summary>
    [JsonPropertyName("features")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public List<string>? Features { get; set; }

    /// <summary>
    /// Number of available operations
    /// </summary>
    [JsonPropertyName("operations_count")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public int? OperationsCount { get; set; }

    /// <summary>
    /// Whether server is MCP-compatible
    /// </summary>
    [JsonPropertyName("mcp_compatible")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? McpCompatible { get; set; }

    /// <summary>
    /// Additional metadata
    /// </summary>
    [JsonPropertyName("metadata")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public JsonElement? Metadata { get; set; }

    /// <summary>
    /// Create new server info
    /// </summary>
    public ServerInfo(string server, string version, string protocol)
    {
        Server = server;
        Version = version;
        Protocol = protocol;
    }

    /// <summary>
    /// Set features (builder pattern)
    /// </summary>
    public ServerInfo WithFeatures(List<string> features)
    {
        Features = features;
        return this;
    }

    /// <summary>
    /// Set operations count (builder pattern)
    /// </summary>
    public ServerInfo WithOperationsCount(int count)
    {
        OperationsCount = count;
        return this;
    }

    /// <summary>
    /// Set MCP compatibility (builder pattern)
    /// </summary>
    public ServerInfo WithMcpCompatible(bool compatible)
    {
        McpCompatible = compatible;
        return this;
    }

    /// <summary>
    /// Set metadata (builder pattern)
    /// </summary>
    public ServerInfo WithMetadata(JsonElement metadata)
    {
        Metadata = metadata;
        return this;
    }
}

